### PR TITLE
TY: fix match ergonomics

### LINKED ISSUE #3225: Wrong type inference related to match ergonomics and `ref`, `mut` keywords
## Environment

* **Intellij-Rust plugin version:** Latest nightly 174d461fbd15caeb20f6967dd7f5206658094a9d
* **Rust toolchain version:** 1.26.0, 1.30.0
* **IDE name and version:** IntelliJ IDEA 2018.3
* **Operating system:** Linux

## Problem description

Match ergonomics is supported since #2457.
But type inference in match patterns has some problems related to match ergonomics and `ref`, `mut` keywords.

This was noticed when trying the code from a real world case https://users.rust-lang.org/t/is-ref-really-a-legacy-pattern/23712 and checking what IntelliJ-Rust thinks about the code.

Examples:
```rust
struct S { a: i32 }
fn main() {
    let S { a: mut a } = &S { a: 1 };
    a
} //^ Rustc: i32, IntelliJ: &i32
```

```rust
struct S { a: i32 }
fn main() {
    let S { a: ref mut a } = &S { a: 1 };
    a
} //^ Rustc: &mut i32, IntelliJ: &i32
```

```rust
fn main() {
    let (_, &a) = &(1i32, &2i32);
    a
} //^ Rustc: i32, IntelliJ: &i32
```

### LINKED ISSUE #3776: False positive exhaustiveness check
## Environment

* **Intellij-Rust plugin version:** Screenshot from 2019-04-30 12-05-43
* **Rust toolchain version:** Unrelated
* **IDE name and version:** Screenshot from 2019-04-30 12-05-27
* **Operating system:** Ubuntu 18.04 (I believe it's unrelated)

## Problem description
Given this minimum reproducible code:

```rust
enum Marisa {
    Ora,
    Muda,
    Dura,
}

fn main() {
    let _ = match (Marisa::Dura, Some(1_919_810)) {
        (Marisa::Ora, None) | (Marisa::Dura, Some(_)) => 23,
        (Marisa::Muda, None) | (Marisa::Muda, Some(_)) => 233,
        (Marisa::Ora, Some(b)) => b,
        (Marisa::Dura, None) => 233,
    };
}
```

IntelliJ Rust lies that it's not exhaustive:

Screenshot from 2019-04-30 12-07-04

### LINKED ISSUE #3791: Pattern matching on &str false positive
## Environment

Same as #3776

## Problem description
Given this minimum-reproducible-example:

```rust
enum Bla<'a> {
    Rua(&'a str)
}
fn van(bla: &Bla) {
    match bla {
        Bla::Rua("shirasawa") => (),
        Bla::Rua("marisa") => (),
        _ => unreachable!()
    }
}
```

This code compiles fine, but IntelliJ-Rust complains with:
Screenshot from 2019-05-05 00-00-03

## Steps to reproduce

Open any project and put the code given to any file included in the project

### PR DESCRIPTION
**Pull request by vlad20012**
Fixes #3225, fixes #3791, fixes #3776

### CODE REVIEW

File: src/test/kotlin/org/rust/lang/core/type/RsPatternMatchingTest.kt [hunk 1]
```diff
@@ -422,27 +420,27 @@ class RsPatternMatchingTest : RsTypificationTestBase() {
...
             let E::A(mut a) = &E::A(1);
             a;
         } //^ i32
     """)
-    }
 
+    // TODO
+    // TODO
+    // TODO
+    // TODO
```
Comment (Undin) [inline src/test/kotlin/org/rust/lang/core/type/RsPatternMatchingTest.kt]: 4 `TODO`s? why?

Comment (vlad20012) [reply src/test/kotlin/org/rust/lang/core/type/RsPatternMatchingTest.kt]: Oh, I added them not to miss it before commit, but I missed it anyway :(

**Pull request merged**